### PR TITLE
Fix: multiple groups of events for same course

### PR DESCRIPTION
**Pull request by Thyroidr**
Added two new functions in `/src/utils.js` to allow appending to url query parameters and deleting a specific value. Used these two new functions in `/src/App.js` in `changeOccurrences`.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -69,6 +69,30 @@ export const setQueryParam = (param, value) => {
   localStorage.savedQueryParams = qs.toString()
 }
 
+export const appendQueryParam = (param, value) => {
+  const qs = new URLSearchParams(window.location.search)
+  if (qs.get(param) && value) {
```
Comment (CraftyDH) [inline src/utils.js:74]: Any reason for this if statement? From what I can see it still works perfectly fine with just L75.

File: src/utils.js [hunk 2]
```diff
@@ -69,6 +69,30 @@ export const setQueryParam = (param, value) => {
...
+  } else {
+    qs.set(param, value ?? qs.get(param) ?? '')
+  }
+  window.history.replaceState(null, '', '?' + qs.toString())
+  localStorage.savedQueryParams = qs.toString()
+}
+
+export const deleteQueryParam = (param, value) => {
+  const qs = new URLSearchParams(window.location.search)
+  if (qs.get(param) && value) {
```
Comment (CraftyDH) [inline src/utils.js:85]: Same as above. It seems to work and makes more sense, with just the if branch.

**Review by CraftyDH — CHANGES_REQUESTED** [review]
Not sure but I believe that these if branches are redundant.

Otherwise LGTM.

File: src/utils.js (hunk 1 above)
Comment (Thyroidr) [reply src/utils.js:74]: Just to copy the discord discussion here for future reference: removing this `if` gives an extra comma in front of the query string from the first append.

Comment (CraftyDH) [reply src/utils.js:74]: As discussed in discord, we don't need to check for empty value.

**Review by CraftyDH — APPROVED** [review]

File: src/utils.js [hunk 3]
```diff
@@ -69,6 +69,27 @@ export const setQueryParam = (param, value) => {
...
+export const appendQueryParam = (param, value) => {
+  const qs = new URLSearchParams(window.location.search)
+  if (qs.get(param)) {
+    qs.set(param, [...new Set([...qs.get(param).split(','), value])].join(',')) // remove duplication
+  } else {
+    qs.set(param, value)
+  }
+  window.history.replaceState(null, '', '?' + qs.toString())
+  localStorage.savedQueryParams = qs.toString()
+}
```
Comment (OliverBalfour) [inline src/utils.js:81]: Could we write this in terms of `setQueryParam`?
Something like:
```js
export const appendQueryParam = (param, value) => {
  const existing = (qs.get(param) ?? []).split(',')
  setQueryParam(param, [...new Set([...existing, value])].join(','))
}
```

File: src/utils.js [hunk 4]
```diff
@@ -69,6 +69,27 @@ export const setQueryParam = (param, value) => {
...
+
+// delete `value` entry from the specified parameter if it exists
+export const deleteQueryParam = (param, value) => {
```
Comment (OliverBalfour) [inline src/utils.js:84]: This name sounds like it's deleting the entire query parameter, not an element of a comma separated list. Could we name it `popQueryParamElement` and `appendQueryParamElement` or similar?

Likewise, could we rewrite it using `setQueryParam`? eg (untested):
```js
export const popQueryParamElement = (param, value) => {
  const existing = new Set[...(qs.get(param) ?? []).split(',')])
  existing.delete(value)
  setQueryParam(param, [...existing].join(','))
}
```

**Review by OliverBalfour — COMMENTED** [review]
Thanks for doing this! I think we can simplify the code a little but looks good otherwise

File: src/utils.js (hunk 3 above)
Comment (Thyroidr) [reply src/utils.js]: This leaves a comma in front of the value, rewrote to take that into account

File: src/utils.js [hunk 5]
```diff
@@ -69,6 +69,21 @@ export const setQueryParam = (param, value) => {
...
+  const qs = new URLSearchParams(window.location.search)
+  setQueryParam(param, qs.get(param) ?
+    [...new Set([...qs.get(param).split(','), value])].join(',') :
+    value)
+}
+
+// delete `value` entry from the specified parameter if it exists
+export const popQueryParamElement = (param, value) => {
+  const qs = new URLSearchParams(window.location.search)
+  const existing = new Set([...(qs.get(param) ?? []).split(',')])
```
Comment (pl4nty) [inline src/utils.js:82]:
```suggestion
  const existing = new Set(qs.get(param)?.split(',') ?? [])
```

File: src/utils.js [hunk 6]
```diff
@@ -69,6 +69,21 @@ export const setQueryParam = (param, value) => {
   localStorage.savedQueryParams = qs.toString()
 }
 
+export const appendQueryParamElement = (param, value) => {
+  const qs = new URLSearchParams(window.location.search)
+  setQueryParam(param, qs.get(param) ?
```
Comment (pl4nty) [inline src/utils.js:74]:
```suggestion
  setQueryParam(param, [...new Set(qs.get(param)?.split(',') ?? []).add(value)].join(',')
```

**Review by pl4nty — COMMENTED** [review]
Just some small readability fixes, I'm on mobile tho so probably some typos - wouldn't recommend committing them directly.

Otherwise LGTM, thanks for the contribution!

File: src/utils.js [hunk 7]
```diff
@@ -69,6 +69,21 @@ export const setQueryParam = (param, value) => {
...
+    [...new Set([...qs.get(param).split(','), value])].join(',') :
+    value)
```
Comment (pl4nty) [inline src/utils.js:76]:
```suggestion
  setQueryParam(param, [...new Set(qs.get(param)?.split(',') ?? []).add(value)].join(',')
```

Comment (Thyroidr) [reply src/utils.js]: Kind of annoying that this does not work (leaves a comma in front) as the optional chaining operator only cares about `null` and `undefined`, but not empty string...

Comment (pl4nty) [reply src/utils.js]: ahhh split changes `''` to `['']`, rip

Comment (pl4nty) [reply src/utils.js]: how about

```javascript
setQueryParam(param, [...new Set((qs.get(param) || null)?.split(',')).add(value)].join(',')
```

this doesn't matter at all lol but I like playing with the new js features

Comment (Thyroidr) [reply src/utils.js]: Yep made that change and I think it works

**Pull request merged**